### DISCUSSION TITLE
How do I correctly reference a database context in one project from an API in another project, without duplicating the ConnectionString configuration?

### DISCUSSION THREAD

**Discussion by Xen0byte** [microsoft/aspire#1525] [Q&A]
I've been following the eShopLite example but I can't figure out what is going on because when I try to inject my database context I keep getting `"ConnectionString is missing. It should be provided in 'ConnectionStrings:database' or under the 'ConnectionString' key in 'Aspire:Microsoft:EntityFrameworkCore:SqlServer' or 'Aspire:Microsoft:EntityFrameworkCore:SqlServer:DatabaseContext' configuration section."`. This is (roughly) my setup.

I've got a database manager project that registers a SQL Server service...
`builder.AddSqlServerDbContext<MyContext>("MyConnectionStringName");`
...and everything works fine in that project.

The Aspire.AppHost builder looks like something along the lines of...
```C#
IResourceBuilder<ProjectResource> database = builder.AddProject<My_Database_Manager>("My Database");
builder.AddProject<My_WebPortal_API>("My Web Portal API").WithReference(database);
```

But if I register `MyContext` as a dependency in my API project and try to inject it into a controller it doesn't quite work, and I'm not sure what I'm doing wrong. I was expecting that if I just do `builder.AddSqlServerDbContext<MyContext>("MyConnectionStringName");` in my API project then there is some magic that happens behind the scenes and the connection string is retrieved from some shared Aspire context.

In the eShopLite sample solution I noticed that one project had some JSON config along the lines of...
```JSON
"ConnectionStrings": {
  "catalogdb": "Run this project via the AppHost to ensure this value is populated with a valid connection string"
  }
```
...so I tried a placeholder connection string in my API project too, but that didn't work either.

Could someone please point me in the right direction, because I'm not sure what I'm doing wrong. I just want to reference my database context from my API project without duplicating the connection string. Is this possible?

**Comment by davidfowl** [comment]
Can you show the entire AppHost project's Program.cs and the application's Program.cs (where the DbContext wire up is happening?)

**Comment by Xen0byte** [comment]
Hi @davidfowl, the AppHost is https://github.com/Project-KONGOR-Open-Source/ASPIRE/blob/main/ASPIRE.AppHost/ASPIRE.cs and the database project is https://github.com/Project-KONGOR-Open-Source/ASPIRE/blob/main/MERRICK.Database/MERRICK.cs. Essentially, what I'm trying to achieve is to give https://github.com/Project-KONGOR-Open-Source/ASPIRE/blob/main/ZORGATH.WebPortal.API/ZORGATH.cs access to the database context without duplicating the connection string in the configuration, if that is possible.

**Reply by davidfowl** [reply]
The only way to avoid the duplication would be to specify the connection string via the app host.

**Reply by Xen0byte** [reply]
Thank you very much, I will give it a try and come back with the results.

**Reply by Xen0byte** [reply]
Hi @davidfowl, as I keep trying different things out I keep thinking that maybe my question is maybe somewhat invalid because, unless there is something that I'm missing, one copy of the connection string need to live inside the database project for migrations etc. to work, so even if all the other project get the connection string from AppHost, I think this means that there will still be a minimum of two copies of it in the solution. That being said, simply specifying the connection string via the AppHost does not seem to work for me, and I'm still getting a `ConnectionString is missing. It should be provided in 'ConnectionStrings:MyConnectionName' or under the 'ConnectionString' key in 'Aspire:Microsoft:EntityFrameworkCore:SqlServer' or 'Aspire:Microsoft:EntityFrameworkCore:SqlServer:MyDatabaseContext' configuration section.` exception.

Going a different approach and doing `builder.AddSqlServer("MyConnectionName")` and using this resource as a reference kind of works but it is overriding my connection string on development and spinning a container up instead, and there doesn't seem to be a way around this.

**Comment by Xen0byte** [comment]
Also, thank you very much for the prompt response!

**Comment by UrielMhezzek** [comment]
I don't know how to do it with Aspiere, but you can add a key vault and store the connection string there. The key vault can be added to the Aspiere host and injected into the other applications.

**Reply by Xen0byte** [reply]
Thank you for the reply @UrielMhezzek, you are completely right, and there are many ways to do it, I'm just trying to figure out what the correct way is to achieve this with Aspire, because it feels like if you create a reference to a database project you should at least have access to the connection string from the project that references it. It almost feels like you should not even need to explicitly declare a database dependency (e.g. builder.AddSqlServerDbContext<SomeType>() ...) from a project that references a database project, but that's probably the topic of another discussion.
